controllers/tasks.js

```diff
@@ -109,68 +109,41 @@ const createTask = [
     [task] = task.ops;
 
     // fetch *whole* space tree
-    logger.debug('fetching whole space tree');
     const spaceTree = await fetchWholeTree(itemsCollection, [
       ObjectId(task.spaceId),
     ]);
 
     // fetch actions cursor of retrieved tree
     const spaceIds = spaceTree.map((space) => space.id);
-    logger.debug('fetching actions cursor');
     const actionsCursor = await fetchActions(actionsCollection, spaceIds);
 
-    // create file name to write data to
-    logger.debug('beginning file write');
+    // create file name to write data to; create metadata object; write/upload/hide file
     const fileName = `${task.createdAt.toISOString()}-${task._id.toString()}.json`;
-
-    // write opening array bracket onto JSON file
-    await fs.writeFile(fileName, '[\n', (err) => {
-      if (err) {
-        logger.error(err);
-      }
-    });
-
-    // iterate over cursor and write each document onto JSON file
-    let separator = '';
-    await actionsCursor.forEach((document) => {
-      fs.appendFile(
-        fileName,
-        separator + JSON.stringify(document, null, 2),
-        (err) => {
-          if (err) {
-            logger.error(err);
-          }
-        },
-      );
-      if (!separator) {
-        separator = ',\n';
-      }
-    });
-
-    // write closing bracket onto JSON file
-    await fs.appendFile(fileName, '\n]\n', async (err) => {
-      if (err) {
-        logger.error(err);
-      }
-    });
-
-    // trigger upload/delete/hide/task complete operations
-    try {
-      logger.debug('file write complete; uploading file');
-      const fileId = await uploadFile(
-        // for testing, use spaceId 5ed5f92233c8a33f3d5d87a5
+    const metadata = {
+      spaceTree,
+      createdAt: new Date(Date.now()),
+    };
+    writeDataFile(fileName, actionsCursor, metadata, () => {
+      uploadFile(
         `https://graasp.eu/spaces/${task.spaceId}/file-upload`,
         req.headers.cookie,
         fileName,
-      );
-      logger.debug('hiding file');
-      await hideFile('https://graasp.eu/items/', req.headers.cookie, fileId);
-      logger.debug('marking task complete');
-      await markTaskComplete(tasksCollection, task._id, fileId);
-    } catch (err) {
-      logger.error(err);
-      await tasksCollection.deleteOne({ _id: task._id });
-    }
+      )
+        .catch(async (err) => {
+          logger.error(err);
+          logger.debug('operation failed during file upload');
+          await tasksCollection.deleteOne({ _id: task._id });
+          logger.debug('attempting to delete created resource');
+          deleteFileLocally(fileName);
```

Comment:
Also here you need to throw `err`, otherwise the next then() will still execute.

Reply:
👍